[STEP3] 자동차 경주

makga87 opened the pull request:
~~자동차 경주 STEP3 개발 진행~~
~~테스트 코드는 작은 단위로 진행~~
~~Car : 자동차 명과 위치 값을 가진 객체~~
~~RaceCondition : 경주를 위한 준비 조건들 (자동차가 준비되어야 하고, 이동에 대한 조건이 준비)~~
~~CarRace : Race인터페이스의 구현체~~
~~MonitorView : 자동차 경주 상황을 실시간 모니터링 하는 화면용 객체~~


### 적혀진 리뷰내용 기준으로 현재 로직을 맞추려다 보니, 부적합하다는 판단이 들어, 기존 패키지 구조나 작업 내용을 전체적으로 수정

makga87 committed:
Update README.md

pjhsk113 committed:
feat: README

makga87 committed:
[FEAT] Race 인터페이스 생성

makga87 committed:
[FEAT] Race 도메인 정의 (Entity)

makga87 committed:
[FEAT] Race 뷰 생성

makga87 committed:
[FEAT] Race 상속 받는 CarRace생성

makga87 committed:
[REFACTOR] Race 인터페이스 서비스로 이동

makga87 committed:
[REFACTOR] 라인분리는 비즈니스로직으로 이동

makga87 committed:
[REFACTOR] 테스트 코드 작성을 위한, 내부 로직 파라메터로 받도록 수정

makga87 committed:
[FIX] 라인 체크 두번씩 되는 이슈 수정

makga87 committed:
[TEST] 각 테스트 코드 추가

makga87 committed:
[FEAT] 메인 메서드

makga87 committed:
[FIX] 푸시전 README 충돌 수정

makga87 committed:
[FIX] 입력 예외 조건 추가

ldonghee commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,23 @@
+        int moveCondition = 4;
+
+        Scanner input = new Scanner(System.in);
+        System.out.println("자동차 대수는 몇 대 인가요?");
```
UI 로직을 InputView, ResultView 와 같은 클래스 분리해서 진행해보는건 어떨까요?
클리스를 분리한다면 Race Interface 에서도 활용할 수 있을 것 같아 관리 측면에서 좀 더 나을 것같은 생각이 듭니다:)

makga87 replied:
관련내용 수정 반영했습니다. 뷰는 서비스나 도메인과는 분리가 되어야할 것 같아서, 인터페이스에 넣진 않고, 서비스 로직상에서 호출하도록 개발수정했습니다.

ldonghee commented on src/main/java/service/Race.java:
```diff
@@ -0,0 +1,19 @@
+
+    default void start(int raceTryCount) {
+
+		if(raceTryCount < 1) throw new IllegalArgumentException("시도 횟수는 0이상이어야 합니다.");
```
인덴트 부분이 안맞는 것 같습니다:)

makga87 replied:
해당 부분은 한 줄로 되어 있는데 혹시 어떻게 되어야 하는걸까요?

makga87 replied:
위 내용의 예외처리 부분은, CarRaceStarView의 결과 값인, RaceConditionInput으로  객체로 이동 시켰습니다.
그리고 indent가 안맞는다는 부분이 정확히 어떤 것일까요? 예외가 다음라인으로 내려가야한다는 부분일까요?

pjhsk113 commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,23 @@
+
+    public static void main(String[] args) {
+
+        int moveCondition = 4;
```
> 전진하는 조건은 0에서 9 사이에서 random 값을 구한 후 random 값이 4이상일 경우이다.

위 조건은 자동차 경주 게임에 명확히 정해진 규칙입니다.
이 규칙을 ENUM 상수로 분리해서 따로 관리해 보는건 어떨까요?

makga87 replied:
그렇군요, 경주 시작전 입력된 값들은 걍 main뒀었네요, 관련 내용으로 변경해볼게요

makga87 replied:
> > 전진하는 조건은 0에서 9 사이에서 random 값을 구한 후 random 값이 4이상일 경우이다.
> 
> 위 조건은 자동차 경주 게임에 명확히 정해진 규칙입니다. 이 규칙을 ENUM 상수로 분리해서 따로 관리해 보는건 어떨까요?

해당 내용을 덩그러니 값 하나 있는 ENUM을 만들자니, 적합해 보이지 않고, 적당한용어도 생각나지 않아, '난이도'라는 개념으로 생각하고 ENUM 생생, 난이도별 전진 기준값이 다르도록 설정하였습니다.

pjhsk113 commented on src/main/java/service/domain/RaceCondition.java:
```diff
@@ -0,0 +1,37 @@
+	public RaceCondition(int carCount, int moveCondition) {
+		initCarList(carCount);
+		this.moveCondition = moveCondition;
+	}
```
생성자를 public으로 열어두셨네요!
정적 팩터리 메서드를 사용한다면 생성자는 private으로 두는것이 좋다고 생각합니다.

``` java 
new  RaceCondition(1, 4);
```
형태로 생성이 가능해지면 validation 체크를 우회할 수 있게 되고 
상속이 가능해지면서 의도치않은 클래스가 만들어질 여지를 제공하게 됩니다.

또한, 생성자는 생성의 책임만 가지는게 좋다고 생각합니당.
initCarList() 메서드는 정적 팩터리 메서드로 옮기고 전달받는 형태로 변경해보면 어떨까요?
``` diff
public RaceCondition(List<Car> cars, int moveCondition) {
-       initCarList(carCount);
+       this.carList = cars;
        this.moveCondition = moveCondition;
}
```

makga87 replied:
관련내용 공감이 되네요 수정하겠습니다.

makga87 replied:
> 생성자를 public으로 열어두셨네요! 정적 팩터리 메서드를 사용한다면 생성자는 private으로 두는것이 좋다고 생각합니다.
> 
> ```java
> new  RaceCondition(1, 4);
> ```
> 
> 형태로 생성이 가능해지면 validation 체크를 우회할 수 있게 되고 상속이 가능해지면서 의도치않은 클래스가 만들어질 여지를 제공하게 됩니다.
> 
> 또한, 생성자는 생성의 책임만 가지는게 좋다고 생각합니당. initCarList() 메서드는 정적 팩터리 메서드로 옮기고 전달받는 형태로 변경해보면 어떨까요?
> 
> ```diff
> public RaceCondition(List<Car> cars, int moveCondition) {
> -       initCarList(carCount);
> +       this.carList = cars;
>         this.moveCondition = moveCondition;
> }
> ```

RaceCondition은 이제 사라지고, RaceConditionInput으로 비슷한 용어지만 View에 쓰이는 객체로 변경 되었습니다. 
위 리뷰에 대한 반영은 CarRace 객체에 녹였습니다.

pjhsk113 commented on src/main/java/service/domain/RaceCondition.java:
```diff
@@ -0,0 +1,37 @@
+
+public class RaceCondition {
+
+	private final ArrayList<Car> carList = new ArrayList<>();
```
ArrayList<> 같은 구현체를 타입으로 가져가는 것보다 
List<>와 같은 인터페이스를 타입으로 주고 구현체는 조금 더 유연하게 가져가는게 좋을 것 같습니다. :)

makga87 replied:
넵, 수정하겠습니다.

makga87 replied:
CarRace객체로 이관되며, 해당 내용 적용했습니다.

pjhsk113 commented on src/main/java/service/domain/Car.java:
```diff
@@ -0,0 +1,20 @@
+    public Car(String carName, int poisition) {
+        this.carName = carName;
+        this.poisition = poisition;
+    }
```
자동차(Car) 객체가 이름과 자신의 위치로 자신을 표현하고
이동이라는 한 가지의 책임만가지고 있네요!
아주 명확한 클래스 분리인 것 같습니다.👍👍👍

HwangSeonHo replied:
이후 확장을 가정 했을 때 자동차는 충돌 주유 수리 등등 많은 액션을 가지게 될 수 있고
이에 따라 메소드가 많이 늘어나게 되면 Car객체가 커져서 기능을 찾기 힘들 것 같습니다.
한가지 행동을 하되 행동도 추상화하면 이상할까요?

pjhsk113 replied:
이후에 말씀하신 요구사항이 생긴다면 그때 추상화해도 될 것 같아요~!
미래에 일어날 수 있는 혹은 일어나지 않을 수 있는 일까지 예상해서 지금 당장 추상화 해 놓는건
오버엔지니어링에 가깝지 않을까요!?

제 생각엔 지금 집중해야 할 일은 기능이나 요구사항이 확장되었을 때 유연하게 대처할 수 있는 
객체 구조 설계가 우선적으로 고려되어야 한다고 생각합니다.

HwangSeonHo replied:
미래에 일어날 수 있는 혹은 일어나지 않을 수 있는 일을 예상하는 것이 
기능이나 요구사항의 확장을 대처하는 내용이라고 생각이 들었습니다!

pjhsk113 commented on src/main/java/service/Race.java:
```diff
@@ -0,0 +1,19 @@
+package service;
+
+public interface Race {
```
기능적 측면에서 Race라는 인터페이스를 분리하신 이유가 궁금합니다!
확장 가능성을 염두하시고 만드신걸까요?

makga87 replied:
넵 추상화 과정에서 나온 인터페이스입니다. 여러 종류의 경주가 있고, 그 중 자동차 경주를 개발한 느낌으로 진행해봤습니다.

pjhsk113 commented on src/main/java/service/Race.java:
```diff
@@ -0,0 +1,19 @@
+
+    default void start(int raceTryCount) {
+
+		if(raceTryCount < 1) throw new IllegalArgumentException("시도 횟수는 0이상이어야 합니다.");
```
if문에 항상 중괄호를 사용하는 것이 Java Code Convention입니다!

[https://google.github.io/styleguide/javaguide.html](https://google.github.io/styleguide/javaguide.html)

pjhsk113 commented on src/main/java/service/domain/RaceCondition.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.util.ArrayList;
+
+public class RaceCondition {
```
RaceCondition이라는 이름만 봤을 때는 레이싱 조건에 대한 역할을 수행하는 클래스라고 예상했는데,
실제 하고 있는 일은 여러대의 자동차를 생성하거나, 이동을 명령하는 역할을 수행하고 있네요!

여러대의 자동차를 생성하고 관리하고 있으니, 더 명확한 클래스명을 고민해보면 좋을 것 같습니다.
또, 자동차 객체의 리스트를 관리하고 있으므로 Car 객체 리스트에 대한 책임을 가지는 **일급 컬렉션**으로 변경해보는 건 어떨까요?

makga87 replied:
이 부분은 메서드로 수정할 예정입니다. 처음 의도는 자동차 경주를 위한, 준비항목들이란 개념으로 만들었습니다.

makga87 replied:
관련 내용은 CarRace객체에 수정하였습니다.

pjhsk113 commented on src/main/java/service/domain/RaceCondition.java:
```diff
@@ -0,0 +1,37 @@
+
+	public ArrayList<Car> getCarList() {
+		return this.carList;
+	}
```
컬렉션을 반환하게되면서 자신의 멤버를 노출하게 되었네요.
이렇게 되면 클라이언트쪽에서 carList에 Car객체를 추가할 수 있게되고 해당 객체의 불변성이 깨지게 됩니다.
만약 내부 컬렉션이 외부에서 필요하다면 Stream 객체를 반환해주는 건 어떨까요?
스트림은 처리하는 데이터소스를 변경하지 않습니다. 즉, 원본 데이터를 훼손하지 않기 때문에 더 안전하다고 볼 수 있습니다.
```java
public Stream<Car> getCarList() {
    return carList.stream();
}
```

makga87 replied:
네, 좋은 아이디어 같습니다.

makga87 replied:
관련 내용은 CarRace객체에 수정하였습니다.

pjhsk113 commented on src/test/java/service/domain/RaceConditionTest.java:
```diff
@@ -0,0 +1,23 @@
+
+        assertEquals(randomValue >= moveCondition, raceCondition.isAdvanceCondition(randomValue));
+
+    }
```
자동차의 전진 여부의 올바른 테스트 값 검증은
각 레이스 별, 랜덤 값에 의해 Car 객체가 전진했냐 안했냐가 되어야하지 않을까요?!
해당 테스트는 단순히 랜덤 값이 4 이상인지, 이하인지를 테스트하는 코드같네요!

랜덤 값으로 생성되는 입력 값을 기준으로 자동차가 전진했는지, 전진하지 않았는지를 
손쉽게 테스트할 수 있는 방법을 고민해보시면 좋을 것 같습니다!!

makga87 replied:
자동차의 전진 여부는, 말그대로, '여''부'만 테스트만 진행합니다. 진행은 Car.move로 테스트로 진행했습니다.
여부에 따라, 자동차 전진을 하니, 각각의 단위 테스트만 진행을 한 내용입니다. 해당 메서드의 조립과 로직 진행 흐름은 비즈니스로직의 테스트이므로 현 테스트보다 좀더 큰 단위라 생각하고, 해당 테스트 로직은 작성안했습니다.

makga87 replied:
해당 부분은 일단 CarRaceDifficulty.isMoveOk로 변경 했습니다. 랜덤 값이 지정한 난이도에 이동 조건이 맞는지 확인하는 로직입니다.

pjhsk113 commented on src/main/java/service/CarRace.java:
```diff
@@ -0,0 +1,33 @@
+            String result = view.render(car);
+            System.out.println(result);
+        });
+    }
```
> UI 로직을 InputView, ResultView와 같은 클래스를 추가해 분리한다.

현재 결과 출력을 위한 로직들이 여러군대 흩어져 있는 느낌인데요. 
위 규칙을 적용해서 ResultView를 생성하고 모든 출력 로직을 위임해서 응집도를 높여보는 건 어떨까요!?

makga87 replied:
입력값을 받는 부분이 InputView로 빠지는건가요? 우선 시스템입력 값이기 때문에, 메인 메서드로 작성했으나, 한 번 입력 뷰로 변경에 도전해보겠습니다.
ResultView의 해당하는 부분은 MonitorView라 생각합니다. 반복처리를 다 한 후, 뷰를 출력하는 것이 아닌, 반복 루프 1회를 결과뷰로 생각하여 그리 작성했습니다.

makga87 replied:
우선 Start, Result view 로 나누어서 진행 해보았습니다. 분산된 뷰를 모으기 위해 전체적인 소스수정이 이루어졌으니, 새로 확인 바랍니다.

pjhsk113 commented on src/main/java/service/CarRace.java:
```diff
@@ -0,0 +1,33 @@
+            Random random = new Random();
+            if (raceCondition.isAdvanceCondition(random.nextInt(10))) car.move();
+        });
+    }
```
해당 메서드는 indent가 2 이상인 것으로 보입니다.
또한, 랜덤 값을 생성하고 자동차의 이동 명령을 하는 두 가지일을 하고 있습니다.

``` java
public void race() {
    raceCondition.moveCars();
}
```
위와 같이 랜덤 값 생성 로직을 분리하고 raceCondition으로 명령 자체를 위임해보는건 어떨까요?

makga87 replied:
해당 내용은 CarRace객체에 수정했고, race와 move 메서드로 나누었습니다.

pjhsk113 reviewed (commented):
형근님~! 스터디 열심히 참여해주셔서 감사합니다.😄
코멘트 몇가지 남겨두었습니다.
혹시 궁금한 점 있으시면 코멘트나 DM 남겨주세요~

HwangSeonHo commented on src/main/java/service/domain/Car.java:
```diff
@@ -0,0 +1,20 @@
+package service.domain;
+
+public class Car {
```
CarRace가 Race를 implement해서 여러 종류의 경주를 추상화했다면
Car 역시 Vehicle 같은 인터페이스로 추상화되는 게 좋을 것 같습니다.

makga87 replied:
좋은 의견입니다. 반영하겠습니다.

makga87 replied:
관련 내용은 추후 적용하겠습니다.

HwangSeonHo commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,23 @@
+        System.out.println("시도할 회수는 몇 회 인가요?");
+        int tryCount = input.nextInt();
+
+        CarRace carRace = new CarRace(RaceCondition.of(carCount, moveCondition));
```
이 부분도 race.start가 될 수 있을듯합니다

makga87 replied:
이 부분은 ready로 받는 부분으로 바꿔볼까 하는데 어떨까요?

makga87 replied:
CarRace를 private으로 숨기고, input view 로부터 받은 파라메터기준으로 CarRacing할 준비가 되었다는 의미의 readyForRace 메서드로 객체를 생성하게 변경하였습니다.

HwangSeonHo commented on src/main/java/service/domain/RaceCondition.java:
```diff
@@ -0,0 +1,37 @@
+
+	private void initCarList(int carCount) {
+		for (int count = 0; count < carCount; count++) {
+			carList.add(new Car("", 0));
```
자동차가 초기화되는 과정 이후에 특정 차량을 참조 할 수 없을 것 같아서
차량 별 고유번호가 있으면 좋을 것 같습니다.

makga87 replied:
해당 부분은 뒤의 STEP4였나요? 해당 요구사항을 참고해보니, 이름 작성하는 부분이 있어 해당 브랜치에서 이름으로 대체 적용 해보겠습니다.

HwangSeonHo replied:
STEP4 내용은 보지 못했는데 차량 경주가 차량 별 순위나 출전 차량 별 정보 검색 같은 내용이 필요 하다고 생각했습니다! 
차량 이름은 Car가 가지기 때문에 조회를 위해 차량 인덱스를 가진 상위 map가 하나 더있으면 좋을 것 같습니다. 차량이름-인덱스로 매칭된 map이라던지 list에서 조회보다는 map으로 인덱스 찾아서 조회하는게 빠를 것 같습니다.
CarList는 ArrayList 자료구조가 사용되었는데 왜 ArrayList 구조가 사용되었는지 궁금합니다 인덱스가 필요 없다면 쓸 수 있는게 더 많아지는 것 같아서요!

https://velog.io/@95qwer/%EC%9E%90%EB%B0%94-%EC%9E%90%EB%A3%8C%EA%B5%AC%EC%A1%B0-%EC%86%8D%EB%8F%84

makga87 replied:
출력 결과물의 요구사항은 결국 생성된 순서대로, 차량의 전진상태를 출력하는 것입니다. 

map으로 하며, 차량의 이름을 가지게 된다면,  1차적으로 화면의 출력시, 출력 순서를 지키기 위한, 차량의 이름을 정렬한 별도의 List가 추가로 필요해보이고요.

map으로만 한다고 해도 역시, 화면에 출력을 할 때에는, keySet을 뽑아서 컬렉션을 루프로 돌리는 형태로 출력을 하는 방식을 쓰게 될 것 같네요

결국 map이나 arraylist나 O(n)은 변함이 없을 것 같고, 로직적으로도 arraylist를 쓰는 편이 더 간단하다고 판단했습니다. (결국엔 클린코드가 최종 목표니까요)

만약, 이번 주제가, 특정 차량만을 '픽'해서 결과를 내야한다면, 저도 map을 사용했을 것 같네요.

개인적인 생각으론 성능에 대한 부분은 비즈니스 로직에서 논하기에는 조금 나중에 해야하지 않을까 합니다. (데이터 스토리지가 아닌 이상?) 성능을 논할 정도의 데이터를 map 혹은 list에 적재를 한다면 그때에는, 오히려 데이터를 분할해서 진행해야하지 않을까요? 
일정 수준의 데이터 양에서는 두 자료구조의 속도차는 느끼기 어려울 것이라 생각이 됩니다. 

비슷한 상황이 자바 api 에서도 존재합니다, 예를 들면, StringTokenizer와 split메서드를 이용해서 문자열을 분리할 때, 내부 로직은 StringTokenizer가 더 비효율적으로 보이나, 특정길이의 문자열과 문자열 타입에서는 오히려 속도가 훨씬 빠르기도 하거든요. 즉, 로직의 비효율이 눈에띄게 나타나려면, 데이터 자체가 몸집이 커야하고, 일반적인 저희의 서버 시스템에서는, 데이터가 큰경우에는 나눠서 요청하는 방식을 차용하기 때문에, 고려를 잘 안 하였던 상황입니다

지극히 개인적인 생각을 좀 적어보았습니다. 

4줄 요약
1. 요구사항이 굳이 map이 필요하지 않은 상황이다.
2. 요구사항이 그렇기에, map을 써도, 로직에서 처리되는 loop는 동일할 것이다
3. 요구사항 기준으론 map이나, list나 성능차이는 거의 못느낄 것 같다
4. 입력값 데이터량이 많다면 한번에 않고, 보통 데이터를 페이징하여 진행하기 때문에, 3번과 같은 상황일 것이다

HwangSeonHo replied:
데이터가 크면 데이터부터 분할을 먼저 고려하면 되는 거네요! 요약까지 감사합니다 👍 👍

makga87 committed:
[REFACTOR] 패키지 구조 변경

makga87 committed:
[REFACTOR] https://github.com/Effective-Tilinkle/racingcar-with-java/pull/4#discussion_r1071636869
전진 조건 값을, Enum의 어떤 값으로 하면 좋을지, 고민, 단순히 상수로 빼기도 굳이라는 애매함이 있음.
난이도라는 명목을 부여하면 어떨까 하여, 해당 컨셉으로 변경 해봄

makga87 committed:
[REFACTOR] https://github.com/Effective-Tilinkle/racingcar-with-java/pull/4#discussion_r1071652400
https://github.com/Effective-Tilinkle/racingcar-with-java/pull/4#discussion_r1071651521
RaceCondition보단 도메인 쪽에서 Car List가 필요해보이고, 전진 조건 역시, 도메인(CarRace)에 해당되는 조건으로 여겨져 삭제

makga87 committed:
[REMOVE] PR 요구사항에 맞추려니 기존 구조가 맞지 않아 수정하면서, 연관 없는 코드들 삭제

makga87 committed:
[REFACTOR] input view, result view 분리 변경

makga87 committed:
[ADD] input view 입력 결과치를 전달하는 객체 생성

makga87 committed:
[REFACTOR] Race 인터페이스 변경

makga87 committed:
[REFACTOR] 난이도 객체에 전진 여부에 대한 로직 추가

makga87 committed:
[REFACTOR] CarRace 내부 로직 변경 진행

makga87 committed:
[REFACTOR] MAIN 메서드를 서비스 클래스에서 진행하도록 수정

makga87 committed:
[REMOVE] 모니터링 뷰 제거

makga87 committed:
[REFACTOR] 메서드명 변경 및 레이스 추적을 위한 시도 번호 파라메터 추가

makga87 committed:
[REFACTOR] Race 인터페이스 변경으로 인한 수정, 레이스 넘버 기준으로 상태값 저장

makga87 committed:
[REFACTOR] 결과 뷰 분리 및, 화면 출력 개발

makga87 committed:
[REFACTOR] 변경된 로직, 메서드명 등의 적용

makga87 committed:
[REFACTOR] InputView 요청에 따라 개발, 입력한 결과 값 전달할 VO 객체 추가

makga87 committed:
[REFACTOR] Main Class는 도메인을 호출 하므로, application service 패키지로 분리

makga87 committed:
[TEST] 테스트 케이스 작성

makga87 committed:
[TEST] 테스트 케이스 만들지 않은것은 지움

makga87 committed:
Merge branch 'master' into step3

pjhsk113 reviewed (commented):
코드를 돌려봤는데 정상적으로 동작하지 않네요🥺
수정 후 다시 PR 올려주세용!!

그리고 스터디 진행 방법을 간략히 정리해뒀습니다.
https://janghopark.notion.site/0f459eb170974a2bb2f7cad680cda7b3
이 방법으로 진행 부탁드립니다!!

pjhsk113 commented on src/test/java/domain/CarRaceTest.java:
```diff
@@ -0,0 +1,24 @@
+		carRace.start(1);
+
+		Stream<Car> cars = carRace.getRaceStatus(1);
+		assertEquals(3, cars.toList().size());
```
toList()는 무엇인가요?! 
static import도 아닌것이....
![image](https://user-images.githubusercontent.com/32737230/213186697-a62eb051-de80-4166-84f1-c550efd86f14.png)
빨간줄이 나오고있어요!!

makga87 replied:
> toList()는 무엇인가요?! static import도 아닌것이.... ![image](https://user-images.githubusercontent.com/32737230/213186697-a62eb051-de80-4166-84f1-c550efd86f14.png) 빨간줄이 나오고있어요!!

자동 완성 기능으로 작성된 것 같은데, 아까 사내에서 해당 테스트를 했었는데, 아마 테스트 후, 뭔가 코드를 수정했었나 보네요
stream count()로 대체 수정했어요~

makga87 committed:
[FIX] https://github.com/Effective-Tilinkle/racingcar-with-java/pull/4/files#r1073551185
테스트 코드 잘못된 메서드 작성된 부분이 있어수정

makga87 committed:
[FIX] 상태값이 상속되는 문제로, Car객체 카피하여 새로운 객체로 생성하는 로직 개발
[REFACTOR] Car 객체, 정적팩토리메서드로 변경

makga87 committed:
[FIX] List<Car> 딥카피 메서드 추가

makga87 committed:
[REFACTOR] Car 객체 정적팩토리메서드 적용으로 코드 수정

makga87 commented on src/main/java/domain/CarRace.java:
```diff
@@ -0,0 +1,65 @@
+            newCars.add(car.copy());
+        }
+        return newCars;
+    }
```
Deep copy 추가

makga87 committed:
[REMOVE] Cloneable 오버라이드 메서드 제거, 관련 throw 제거

makga87 commented on src/main/java/domain/vo/Car.java:
```diff
@@ -0,0 +1,28 @@
+
+    public Car copy() {
+        return new Car(this.carName, this.poisition);
+    }
```
Cloneable을 상속받아 지원 메서드를 썼으나, 여전히 기존 객체 주소를 참고하는 것으로 보여, 직접 구현

ldonghee commented on src/main/java/domain/CarRace.java:
```diff
@@ -0,0 +1,60 @@
+    private static List<Car> createNewRacingCar(int carCount) {
+        List<Car> newCars = new ArrayList<>();
+        for (int cnt = 0; cnt < carCount; cnt++) {
+            newCars.add(Car.of("", 0));
```
[Recommend-Request] 
굳이 바꾸실 필요는 없지만 개인적인 코드 스타일이라 리뷰 남겨봅니다 😃
자동차 이름이 없는 자동차 객체를 반환하는 메서드를 추가해 보는것도 좋지 않을까 싶습니다 ㅎㅎ
메서드 네이밍에 따라 어떤 객체를 반환 되는지 유추 할 수 있어 코드의 가속성 장점이 있다고 생각되어
저는 보통 메서드를 추가하는 편입니다:)

makga87 replied:
아 넵, 해당 내용은 기존엔 필수로 값이 있어야 한다고 생각해서 두 개의 파라메터를 넣도록 개발했으나, STEP3의 요구사항에는 없어, 일단 빈 문자열을 넣었습니다.
STEP4를 미리 본 바로, 해당 인자는 필수 값이 되어야할 것 같아, 이름이 없는 자동차는 생성되지 말아야 하지 않을까 하여 위 방식으로 작성했습니다.

ldonghee commented on src/main/java/view/vo/RaceConditionInput.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    private static void validate(int carCount, int tryCount) {
+        if(carCount < 1) throw new IllegalArgumentException("잘못된 자동차 댓수 입력");
```
[Recommend-Request] 
매직넘버는 상수로 치환해보시는건 어떨까요?

참고해보시면 좋을것 같아 url 전달드립니다.
https://hoonmaro.tistory.com/44

HwangSeonHo replied:
[Question]
보통 단순히 있다 없다를 표현하는 0,1 도 상수로 치환하는게 좋은가요?
상수 치환 값이 특별한 의미를 가진다면 저도 상수 치환이 좋다고 생각하지만 '있다 없다'를 표현하는 이 부분에서 만큼은 워낙 자주 쓰이고 상수명이 들어가는 것 보다 0,1 이 오히려 명확해보인다는 생각이 들어서 궁금하네요

ldonghee replied:
[Good]
이것도 개인적인 코드 스타일이라고 생각됩니다 😃
예를 들어 매직넘버를 상수화하고 이 상수를 여러곳에서 사용한다면 한번의 변경을 통해서 
여러 곳의 수정할 필요가 없어진다는 장점(실수가 줄어듬.. ㅎㅎ)이 있다고 생각되기에 
둘 이상의 매직넘버는 상수화 해서 사용하는 편입니다:)

그리고 단순히 1 이라는 숫자보단  자동차 갯수 제한에 대해 의미를 부여하는 네이밍이 더 명확해 보인다고 생각이 드는데 
이것도 사람마다 차이라 강요는 아닙니다 ㅋㅋ

makga87 replied:
아 넵, 좋은 의견 감사합니다. 
해당 부분은 수정하겠습니다.

pjhsk113 replied:
[Recommend-Must]
if 문에는 항상 중괄호를 사용해주세요~!!

pjhsk113 commented on build.gradle:
```diff
@@ -12,6 +12,8 @@ repositories {
 dependencies {
     testImplementation 'org.assertj:assertj-core:3.22.0'
     testImplementation 'org.junit.jupiter:junit-jupiter:5.8.2'
+    testImplementation 'org.mockito:mockito-core:3.11.2'
```
혹시 mockito 의존성은 왜 넣으신걸까요~??

makga87 replied:
서비스 로직에 대한 테스트를 진행하기 위함 이었습니다.

테스트 코드 작성했다가, 굳이 서비스까지 테스트 하나 싶어서 제거 했는데 딸려 들어갔나보네요

pjhsk113 commented on src/main/java/domain/CarRace.java:
```diff
@@ -0,0 +1,60 @@
+        carRaceStatus.put(tryNo, copyOf(cars).stream());
+    }
+
+    private void move(Car car) {
+        Random random = new Random();
+        if (carRaceDifficulty.isMoveOk(random.nextInt(10))) {
+            car.move();
+        }
+    }
```
[Recommend-Must]
CarRace가 자동차의 전진 여부를 결정하고 있네요..!
이 로직이 Car 객체를 수동적으로 만들고 있는 것 같습니다.

자동차가 스스로 움직임을 제어하는게 더 자연스럽지 않을까요~??
해당 로직은 Car 객체가 가지고 있는게 좋을 것 같습니다 :)

makga87 replied:
자동차 경주에서 조건이 외부 인자가 아닌, 자동차 내의 조건이었나 보군요, 알겠습니다. 수정하겠습니다.

pjhsk113 commented on src/main/java/domain/CarRaceDifficulty.java:
```diff
@@ -0,0 +1,18 @@
+package domain;
+
+public enum CarRaceDifficulty {
+
+    EASY(1),
+    NORMAL(4),
+    HARD(7);
+
+    int moveCondition;
```
[Recommend-Request]
제가 앞서 게임 규칙에 대한 Enum 상수를 만들면 좋겠다고 남겼던 리뷰는
> 전진하는 조건은 0과 9 사이에서 random 값을 구한 후 random 값이 4이상일 경우이다.

위와 같은 게임 규칙에 대한 Enum 상수를 한 곳에서 관리하는 것이 의도였습니다.

예를 들어, 게임 규칙이 다음과 같이 변경된다면 어떨까요?
> 전진하는 조건은 2와 12 사이에서 random 값을 구한 후 random 값이 5이상일 경우이다.

현재 코드에서는 규칙 수정으로 인해 여러 클래스에서 변경이 일어나야 합니다.
즉, 자신의 책임이 아닌 일 때문에 클래스를 수정해야 하는 것이죠.

하지만 게임 규칙만을 담당하는 Enum 상수를 만들고 한 곳에서 관리한다면
규칙이 변경되어도 다른 클래스에는 영향이 가지 않아 변경이 일어나지 않습니다.

다음과 같이 적용해볼 수 있을 것 같아요~!
``` java
public enum GameRules {
    END_INCLUSIVE(9),
    STANDARD_NUMBER(4);

    private final int value;

    GameRules(int value) {
        this.value = value;
    }

    public int getValue() {
        return value;
    }
}
```

``` java
private void move(Car car) {
    Random random = new Random();
    if (carRaceDifficulty.isMoveOk(random.nextInt(GameRules.END_INCLUSIVE.getValue()))) {
        car.move();
    }
}
```

makga87 replied:
[Question] 
만약 위처럼 계속 조건이 바뀌는 케이스라면, Enum 한 곳에서 관리를 한다고 해도 결국, 수정이 발생합니다. 
위 내용에서 하나의 조건만 더 붙인다면, 이렇게 말할 수 있겠네요. "조건이 매번 바뀐다면?"
그럼 결국, Enum 의 상수값을 계속 바꿔야한다는 내용인데, 결국 이러면 일반 Class를 만드는 것이 맞겠네요

게임 규칙이 빈번하게 변경되지 않는다라는 조건이면, 기존 Enum을 그대로 써도 될 것 같습니다.

GameRules이라는 Enum이 서로다른 성격을 가진 정보로 모여 있어보이네요
Enum은 같은 성격으로 모여야하지 않을까요?

pjhsk113 commented on src/main/java/domain/Race.java:
```diff
@@ -0,0 +1,16 @@
+        for (int tryNo = 1; tryNo <= raceTryCount; tryNo++) {
+            race(tryNo);
+        }
+    }
```
불필요한 메서드인 것 같습니다.

https://github.com/Effective-Tilinkle/racingcar-with-java/blob/8ddac8fc658dc43c6077d3ccbcfb1ba162f0f1a4/src/main/java/application/service/CarRaceService.java#L22-L28

위 코드를 아래와 같이 변경해도 똑같이 동작할 것 같네요
``` java  
 for (int tryNo = 1; tryNo <= raceConditionInput.getTryCount(); tryNo++) { 
        race.race(tryNo);
 	View carRaceResultView = CarRaceResultView.of(race.getRaceStatus(tryNo), tryNo); 
 	carRaceResultView.render(); 
 } 
```

makga87 replied:
[Question]
기본적으로 개발 방향은 웹을 main 메서드에 기입하는 것을 차용 했습니다.
저희가 보통 웹서비스 개발을 하면, 서비스 로직을 담당하는 CarRace에서 모든 비즈니스 로직을 완료한 뒤, 그 결과 값을 프론트에서 받아, View로 만들지 않나요?

말씀하신 대로라면, 자동차 경주가 진행되는 중에 View 객체를 계속 호출하는 동작이 될 텐데 의도가 맞으실까요?

해당 케이스라면 굳이 View 객체라기보다, 경주 중에 모니터링이 계속 나와야 하니, CarRace에 resultView 관련 메서드가 생기는게 맞지 않을까 생각이 되는데요?

pjhsk113 commented on src/main/java/domain/CarRace.java:
```diff
@@ -0,0 +1,60 @@
+public class CarRace implements Race<Stream<Car>> {
+
+    private List<Car> cars = new ArrayList<>();
+    private final CarRaceDifficulty carRaceDifficulty;
+    private Map<Integer, Stream<Car>> carRaceStatus = new HashMap<>();
```
[Question]
CarRaceDifficulty를 멤버로 선언하신 이유가 있을까요?

[Recommended-Must]
carRaceStatus는 진행 회차 별 cars의 상태를 담기 위해 선언하신 것으로 이해했습니다.
하나의  `List<Car>`의 상태가 변경되며 Map에 담기다보니 같은 참조가 발생해 deepCopy를 추가하신걸로 보이구요.

하지만 `List<Car>`의 전체 상태를 계속해서 Map에 업데이트할 필요가 있을까요??
저는 List<Car>에 담겨있는 Car 객체의 move() 메서드 호출을 통해 상태값을 변경하는 것만으로도 구현이 가능하다고 생각합니다.

진행 회차에 대한 **출력 구분** 때문에  `List<Car>`를 전체적으로 Map에 업데이트 하는 방법을 생각하셨다면,
각 회차 개념을 Round라는 객체로 표현해보고 round당 모든 Car들을 move()시키고  출력해 주는 메서드를 가지도록
변경해보는 건 어떨까요?

기능 구현보다 도메인을 분석하고 그 안에서 객체를 뽑아내는 연습을 하는게 이 스터디에서 가장 중요한 부분이기 때문에
조금 번거로우시겠지만 개선 요청을 드립니다.🥲🥲

마지막으로, 지난 리뷰에 **일급 컬렉션**이라는 키워드를 잠깐 언급한적 있었는데요!
이 개념은 꼭 알아두셨으면 하는 개념이라 한번 학습해보시고 CarRace 클래스에 적용해주시면 좋을 것 같아요.
https://jojoldu.tistory.com/412

makga87 replied:
> [Question] CarRaceDifficulty를 멤버로 선언하신 이유가 있을까요?
> 
위, CarRace에 move에 대한 조건이 있는 것에 대한 내용이 이어집니다.
자동차의 전진 조건은 외부로부터 제약을 받는 조건으로 생각하고, Car.move를 제어하기 위해 멤버로 선언한 것입니다.

> [Recommended-Must] carRaceStatus는 진행 회차 별 cars의 상태를 담기 위해 선언하신 것으로 이해했습니다. 하나의 `List<Car>`의 상태가 변경되며 Map에 담기다보니 같은 참조가 발생해 deepCopy를 추가하신걸로 보이구요.
> 
> 하지만 `List<Car>`의 전체 상태를 계속해서 Map에 업데이트할 필요가 있을까요?? 저는 List에 담겨있는 Car 객체의 move() 메서드 호출을 통해 상태값을 변경하는 것만으로도 구현이 가능하다고 생각합니다.
> 
> 진행 회차에 대한 **출력 구분** 때문에 `List<Car>`를 전체적으로 Map에 업데이트 하는 방법을 생각하셨다면, 각 회차 개념을 Round라는 객체로 표현해보고 round당 모든 Car들을 move()시키고 출력해 주는 메서드를 가지도록 변경해보는 건 어떨까요?
> 
> 기능 구현보다 도메인을 분석하고 그 안에서 객체를 뽑아내는 연습을 하는게 이 스터디에서 가장 중요한 부분이기 때문에 조금 번거로우시겠지만 개선 요청을 드립니다.🥲🥲
> 
> 마지막으로, 지난 리뷰에 **일급 컬렉션**이라는 키워드를 잠깐 언급한적 있었는데요! 이 개념은 꼭 알아두셨으면 하는 개념이라 한번 학습해보시고 CarRace 클래스에 적용해주시면 좋을 것 같아요. https://jojoldu.tistory.com/412

네, 위 내용에 대한 답변은 이전 문의에 대한 답변과 같습니다.

View와 Service 로직의 분리 시, 웹 개발과 같은 방식으로 이해.
즉, 서비스 로직은 모든 작업을 마친 뒤, 해당 결과 데이터들을 View(Front)로 전달
View에선 해당 데이터들을 화면으로 표시.

이번 브랜치의도가 그게 아니라면, 서비스 로직의 loop 중간에 View를 생성하라는 말씀이신지가 정해지면 좋겠습니다.
(만약 그렇다면, 결국 View 객체로 굳이 분리할 필요가 있는가? CarRace의 진행 중의 메서드의 한 부분으로 느껴지는 점이 있음)

[일급객체관련]
네, 관련 내용은 확인 했었고, 추가적인 객체 없이 충분히, 개발이 가능해서, 굳이 불필요한 로직을 넣는게 필요할까? 해서 작업을 했었습니다.
근데, 지금 다시 보니, 변경해주면 더 명시적으로 보이긴 할 것 같네요

pjhsk113 commented on src/main/java/view/CarRaceStartView.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    public RaceConditionInput render() {
+
+        Scanner input = new Scanner(System.in);
+        System.out.println("자동차 대수는 몇 대 인가요?");
+
+        int carCount = input.nextInt();
+        System.out.println("시도할 회수는 몇 회 인가요?");
+
+        int tryCount = input.nextInt();
```
[Recommend-Request]
`private static final Scanner input = new Scanner(System.in);` 로 선언하고 
각 입력을 메서드로 나눠보는건 어떨까요~?
`inputCarCount() { }`
`inputTryCount() { }`

다른 방법으로는 입력부를 단순히 정적 메서드들을 제공하는 유틸 클래스처럼 사용해도 괜찮을 것 같아요~!
``` java
public static int inputCarCount() { 
    System.out.println("자동차 대수는 몇 대 인가요?");
    return input.nextInt();
}
```

makga87 replied:
[Good]
좋은 방법인것 같네요, 관련 내용으로 수정하겠습니다.

pjhsk113 commented on src/main/java/view/CarRaceResultView.java:
```diff
@@ -0,0 +1,40 @@
+
+public class CarRaceResultView implements View<String> {
+
+	private Stream<Car> cars;
+	private int tryNo;
```
[Recommend-Rquest]
반드시 필요한 멤버일까요~?
출력부를 printRound(cars, tryNo) 혹은 printResult(cars, tryNo)와 같은 이름을 가지는 정적 메서드로 제공하면
단순히 매개변수를 전달받아 처리할 수 있게되어 의존성을 없앨 수 있겠네요!

makga87 replied:
[Quetsion]

CarRace, 결과에 화면 출력에 특화된 View가 되겠군요, 스프링을 생각해서 View, Model등 뭔가 추상화된 인터페이스를 고민해보다가 render로 우선 진행하긴 했었는데, 핀트를 잘 못 잡은 것 같네요. 각 인터페이스들은 제거 후, 각 기능에 특화된 메서드로 변경하면 될까요?

pjhsk113 commented on src/test/java/domain/vo/CarTest.java:
```diff
@@ -0,0 +1,22 @@
+
+	@DisplayName("자동차 이동 시, 포지션 변경을 확인한다")
+	@Test
+	void carMoveTest() {
+		Car car = Car.of("", 0);
+		car.move();
+		assertEquals(1, car.getPoisition());
+
+		car.move();
+		assertEquals(2, car.getPoisition());
+
+	}
```
[Recommend-Request]
@ParameterizedTest를 이용해 더 다양한 케이스를 검증해보면 좋을 것 같아요!

- `@ValueSource`
- `@CsvSource`
- `@MethodSource`

makga87 replied:
[Question]
move에 더 다양한 케이스가 있을까요? @CsvSource 등은 자주 쓰긴 하나, 다양한 테스트 케이스가 필요하지 않다면 잘 안쓰고 있습니다.
필요 케이스를 더 알려주시면 해당 내용으로 수정하겠습니다.

makga87 committed:
[REFACTOR] 일급 컬렉션 생성

makga87 committed:
[REFACTOR] 난수 생성 및 이동 조건 enum 파라메터 변경

makga87 committed:
[REFACTOR] 인터페이스 제거

makga87 committed:
[REFACTOR] 자동차 객체가 외부조건이 아닌 자신이 주체적으로 이동해야한다 하여 해당 방식으로 수정

makga87 committed:
[REFACTOR] 뷰 방식 변경

makga87 committed:
[REFACTOR] Cars 객체, 이동 조건, 인터페이스 제거 적용

makga87 committed:
Merge branch 'step3_refactor' into step3

makga87 committed:
[ADD] CarTest추가

makga87 committed:
[FEAT] Car.move 예외 로직 추가

makga87 committed:
[ADD] Car.move 예외 로직 테스트 추가

pjhsk113 commented on src/main/java/domain/CarRace.java:
```diff
@@ -0,0 +1,39 @@
+public class CarRace {
+
+	private Cars cars;
+	private final CarRaceDifficulty carRaceDifficulty;
```
[Question]
Enum 상수를 사용하는 부분에 `CarRaceDifficulty.Normal.get()` 형식으로 사용해도 될 것 같은데
멤버로 선언하신 이유가 따로 있을까요~??

makga87 replied:
> [Question] Enum 상수를 사용하는 부분에 `CarRaceDifficulty.Normal.get()` 형식으로 사용해도 될 것 같은데 멤버로 선언하신 이유가 따로 있을까요~??

CarRace 객체 생성 시를 의미하시는건가요?
그게 맞다면 의도는 난이도enum에 이동조건과, 랜덤값 범위 조건 2개가 있어서, 객체 생성시 필요한 argument 갯수를 줄이려고, enum을 통째로 넣었습니다.

pjhsk113 commented on src/main/java/domain/CarRace.java:
```diff
@@ -0,0 +1,39 @@
+	private Cars cars;
+	private final CarRaceDifficulty carRaceDifficulty;
+
+	public static CarRace readyForRace(int carCount, CarRaceDifficulty carRaceDifficulty) {
+		return new CarRace(carCount, carRaceDifficulty);
+	}
+
+	private CarRace(int carCount, CarRaceDifficulty carRaceDifficulty) {
+		this.cars = createNewRacingCars(carCount);
+		this.carRaceDifficulty = carRaceDifficulty;
+	}
```
[Recommend-Must]
1. static 변수 -> 인스턴스 변수 -> 생성자 -> 메서드 순으로 작성하는게 자바 코드 컨벤션인 것으로 알고있습니다!
생성자의 위치와 정적 팩터리 메서드의 위치를 바꿔주면 좋을 것 같습니다.

2. 현재 생성자와 정적팩터리 메서드의 역할이 서로 바뀌어 있는 것 같습니다.
다음과 같이 변경하는건 어떨까요!?
```java
private CarRace(Cars cars, CarRaceDifficulty carRaceDifficulty) {
      this.cars = cars;
      this.carRaceDifficulty = carRaceDifficulty;
}

public static CarRace readyForRace(int carCount, CarRaceDifficulty carRaceDifficulty) {
      return new CarRace(createNewRacingCars(carCount), carRaceDifficulty);
}
```

makga87 replied:
> [Recommend-Must]
> 
> 1. static 변수 -> 인스턴스 변수 -> 생성자 -> 메서드 순으로 작성하는게 자바 코드 컨벤션인 것으로 알고있습니다!
>    생성자의 위치와 정적 팩터리 메서드의 위치를 바꿔주면 좋을 것 같습니다.
> 2. 현재 생성자와 정적팩터리 메서드의 역할이 서로 바뀌어 있는 것 같습니다.
>    다음과 같이 변경하는건 어떨까요!?
> 
> ```java
> private CarRace(Cars cars, CarRaceDifficulty carRaceDifficulty) {
>       this.cars = cars;
>       this.carRaceDifficulty = carRaceDifficulty;
> }
> 
> public static CarRace readyForRace(int carCount, CarRaceDifficulty carRaceDifficulty) {
>       return new CarRace(createNewRacingCars(carCount), carRaceDifficulty);
> }
> ```

그렇군요, 생성자는 생성에만 신경을 쓰자는 거군요 이해했습니다. 반영하겠습니다.

pjhsk113 commented on src/main/java/domain/vo/Car.java:
```diff
@@ -0,0 +1,39 @@
+	}
+
+	public void move(CarRaceDifficulty carRaceDifficulty, int randomValue) {
+		if (randomValue <= 0) {
+			throw new IllegalArgumentException("Bad random value input");
+		}
```
[Recommend-Must]
4라는 기준값에 의해 전진 여부를 결정하는데 불필요한 검증 아닐까요~?
음수도 4 이하 값이기 때문에 예외 발생 사유는 아니라고 생각합니다!!
또한, 이 부분 때문에 현재 테스트가 깨지고 있는 것 같습니다!

makga87 replied:
> [Recommend-Must] 4라는 기준값에 의해 전진 여부를 결정하는데 불필요한 검증 아닐까요~? 음수도 4 이하 값이기 때문에 예외 발생 사유는 아니라고 생각합니다!! 또한, 이 부분 때문에 현재 테스트가 깨지고 있는 것 같습니다!

이 부분은 논의가 좀 필요해보입니다, random.nextInt 값으로 나올 값들은 양의 정수인 것으로 압니다만, 위 함수는 random.nextInt 값의 결과를 외부에서 받을 수 있는 함수이고요

그렇기에 기능이 일관되려면, 해당 예외처리가 필요하지 않을까요?

makga87 replied:
추가적으로 테스트가 깨진다는 부분이 어떤 내용일까요? 테스트 코드가 안돌아가는건가요? (전 정상적으로 돌고 있어서 함께 확인해보시죠)

pjhsk113 commented on src/main/java/domain/CarRace.java:
```diff
@@ -0,0 +1,39 @@
+			cars.addCar(Car.of("", 0));
+		}
+		return cars;
+	}
```
[Recommend-Request]
Cars의 생명주기가 생성과 동시에 외부에서 수정할 수 없도록 관리되면 더 좋을 것 같습니다.
다음과 같이 변경해보는건 어떤가요!
```java
private static Cars createNewRacingCars(int carCount) {
    List<Car> cars = new ArrayList<>();
    for (int cnt = 0; cnt < carCount; cnt++) {
	    cars.from(Car.of("", 0));
    }
    return cars;
}
```

makga87 replied:
관련 내용으로 변경 하였습니다.~

pjhsk113 commented on src/main/java/domain/CarRace.java:
```diff
@@ -0,0 +1,39 @@
+		cars.stream()
+			.forEach(car -> {
+				car.move(carRaceDifficulty);
+			});
```
[Recommend-Request]
여러대의 Car를 관리하는게 Cars의 역할이라고 생각하는데요!
해당 로직은 자동차 이동 명령에 대한 로직이니 Cars로 위임하는게 좋을 것 같습니다.
객체들이 구체적인 명령에 의해 수동적으로 동작한다는 느낌보다는
`나는 요청을 보낼테니 처리는 너가 해라`는 느낌으로 진행해보면 각 객체가 수행해야하는 역할이 더 명확하게 보일 것 같습니다!!
```java
public void race() {
    cars.moveCars();
}
```

makga87 replied:
관련 내용으로 수정하였습니다.~!

pjhsk113 commented on src/main/java/domain/vo/Car.java:
```diff
@@ -0,0 +1,39 @@
+		this.poisition = poisition;
+	}
+
+	public void move(CarRaceDifficulty carRaceDifficulty) {
+		Random random = new Random();
+		move(carRaceDifficulty, random.nextInt(carRaceDifficulty.getRandomRange()));
+	}
```
[Recommend-Request]
다음과 같은 형태로 사용해도 되지 않을까요~??!
```java
public void move() {
    Random random = new Random();
    move(carRaceDifficulty, random.nextInt(CarRaceDifficulty.NORMAL.getRandomRange()));
}
```

makga87 replied:
위 내용은 오프라인을 통해 대화를 나눴던 내용

[요약] 
- 자동차 경주 콘솔 게임 형식을 기준으로 개발을 진행해, 난이도는 게임 시작할 때, 외부로부터 입력을 받는 방식으로 개발함

pjhsk113 commented on src/main/java/domain/vo/Cars.java:
```diff
@@ -0,0 +1,42 @@
+
+public class Cars {
+
+	private List<Car> cars;
```
[GOOD]
일급컬렉션 너무 좋습니다.💯💯💯💯💯

pjhsk113 commented on src/main/java/domain/vo/Cars.java:
```diff
@@ -0,0 +1,42 @@
+
+	public void removeCar(int index) {
+		this.cars.remove(index);
+	}
```
[Recommend-Request]
컬렉션의 불변성을 보장하고 상태와 행위를 한 곳에서 관리한다는 컨셉이 일급컬렉션의 장점 중 하나인데요!
일급컬랙션이 가지고 있는 내부 컬랙션의 데이터를 외부에서 조작할 수 있도록 허용한다면 
이를 사용하는 의미가 조금 퇴색됩니다. 🥲
레이싱 게임 라이프 사이클에서도 Cars는 최초 생성 후, 추가되거나 삭제되지 않으므로 불필요한 메서드라고 생각됩니다!!

makga87 replied:
관련 내용들 제거 하였습니다.~

pjhsk113 reviewed (commented):
코드가 전체적으로 파악하기 쉬워진 것 같습니다!!
Request 코멘트는 한 번 확인해보시고 동감되는 부분이 있다면 적용해주시면 좋을 것 같고
Must 코멘트는 수정부탁드립니다!

makga87 committed:
[FIX] https://github.com/Effective-Tilinkle/racingcar-with-java/pull/4#discussion_r1092102587 컬렉션 조작 메서드 제거

makga87 committed:
[FIX] https://github.com/Effective-Tilinkle/racingcar-with-java/pull/4#discussion_r1092093326 cars.move() 로 변경
[FIX] https://github.com/Effective-Tilinkle/racingcar-with-java/pull/4#discussion_r1092081570 cars.from으로 변경
[FIX] https://github.com/Effective-Tilinkle/racingcar-with-java/pull/4#discussion_r1092065140 생성자, 메서드 위치 변경

makga87 committed:
[FIX] https://github.com/Effective-Tilinkle/racingcar-with-java/pull/4#discussion_r1092065140 생성자, 메서드 위치 변경
[FIX] move 함수, randomValue 예외값 조건 부등호 수정

makga87 committed:
[FIX] 스펠링 오탈자 수정

makga87 committed:
[FIX] 테스트 코드 실패, 내용 확인 및 수정

pjhsk113 reviewed (approved):
4따봉 드립니다. 👍👍👍👍
리뷰에 참여해주신 분들께서는 내용 확인 후 approve 부탁드립니다.

HwangSeonHo commented on README.md:
```diff
@@ -1 +1,9 @@
+- 커밋 단위는 작게 유지하며 커밋 메시지는 최대한 자세히 작성한다.
+- 미션 수행을 못했더라도 최소 두 사람 이상의 코드에 리뷰를 하거나 질문(issue)에 답변을 달아야 한다.
+
+- https://janghopark.notion.site/TDD-323df32f45b548a1a9d0b9c8906532a9
```
intellij 리뷰

jinyeeun commented on src/main/java/domain/CarRace.java:
```diff
@@ -0,0 +1,41 @@
+
+public class CarRace {
+
+	private Cars cars;
```
[Good] List<Car> 객체로 선언하지 않고, Cars로 객체를 생성한 것이 좋습니다.

HwangSeonHo commented on src/main/java/domain/vo/Cars.java:
```diff
@@ -0,0 +1,29 @@
+
+public class Cars {
+
+	private List<Car> cars;
```
[Question] 
짜잘한 궁금증입니다.
보통 list로 받을때 변수명 추천으로 cars / carList가 뜨는데
list라면 carList로 쓰는게 타입 추측하기 쉬워서 좋다고 생각했었거든요
요것도 컨벤션이나 가이드 같은게 있나요?

pjhsk113 replied:
정해진 컨벤션은 없는 것으로 알고있습니다. 이건 취향차이인 것 같아요~
내부 컬랙션이 변경되면 클래스명도 같이 변경되야 하기 때문에
저는 특정 컬렉션을 나타내는 네이밍을 잘 안쓰는 편입니다..!!

makga87 replied:
> [Question] 짜잘한 궁금증입니다. 보통 list로 받을때 변수명 추천으로 cars / carList가 뜨는데 list라면 carList로 쓰는게 타입 추측하기 쉬워서 좋다고 생각했었거든요 요것도 컨벤션이나 가이드 같은게 있나요?

특별히 컨벤션때문에 한건 아니고, 저한테는 보통 뒤에 xxxList, xxxs 식으로 두 개가 뜨는 점도 있었고요.
개발하면서 자동차 경주 콘솔게임을 만들 때를 상상하면서 만들다보니, 사용자가 보는 메뉴 기준으로 생각해서 작성하게 되었어요

HwangSeonHo replied:
> 정해진 컨벤션은 없는 것으로 알고있습니다. 이건 취향차이인 것 같아요~ 내부 컬랙션이 변경되면 클래스명도 같이 변경되야 하기 때문에 저는 특정 컬렉션을 나타내는 네이밍을 잘 안쓰는 편입니다..!!

내부 컬렉션이 변경 될 여지가 있다면 List로 안받고 Collection으로 받는 게 더 좋지 않을까요?

-- 아.. map도 있고 이렇게 올라가다간 Object까지 올라가겠네요 말하고보니 이상한것같습니다 ㅋㅋ..

jinyeeun replied:
> 정해진 컨벤션은 없는 것으로 알고있습니다. 이건 취향차이인 것 같아요~ 내부 컬랙션이 변경되면 클래스명도 같이 변경되야 하기 때문에 저는 특정 컬렉션을 나타내는 네이밍을 잘 안쓰는 편입니다..!!

저도 장호님 의견에 동의합니다.
클린코드 - 로버트 C마틴 책에서도 언급되는 내용인데요, 
변수명에 자료타입을 쓰는걸 지양하자고 되어있습니다.

해당 타입이 변경되어도 변수명은 변경하지 않을 수 있기 때문입니다.
carList의 자료형이 List에서 Set이나 Map으로 바뀌었을때 변수명이 변경되지 않으면, 혼란을 야기할 수 있습니다.

HwangSeonHo reviewed (approved):
👍

jinyeeun commented on src/main/java/domain/vo/Car.java:
```diff
@@ -0,0 +1,39 @@
+		move(carRaceDifficulty, random.nextInt(carRaceDifficulty.getRandomRange()));
+	}
+
+	public void move(CarRaceDifficulty carRaceDifficulty, int randomValue) {
+		if (randomValue < 0) {
+			throw new IllegalArgumentException("Bad random value input");
+		}
+		int moveCondition = carRaceDifficulty.getMoveCondition();
+		if (moveCondition <= randomValue) {
+			this.position++;
+		}
+	}
+
```
[Recommend-Request] Car 객체에 게임의 룰이 들어 가 있습니다.
게임의 룰이 변경 될 경우(예를들어, 특정 값 이상이 아니라, 이하로 변경되었을때) Car객체는 변경이 없고, 
차가 움직이는 조건에 대한 클래스가 변경되면 더 좋을 것 같습니다.
오토바이나, 자전거 등 레이싱 가능 한 객체가 늘어났을때 해당 객체들을 수정하는게 아니라, 해당 조건만 수정하면 더 좋을것 같다는 생각이 들었어요.

번외로, Car 객체는 CarRaceDifficulty(게임 난이도)를 몰라도 될 것 같습니다.
CarRaceDifficulty가 들어오는 순간 해당 Car의 객체는 해당 레이싱 게임에서만 사용 가능한 객체가 된것 같습니다.
어떨때 차가 움직여야한다는 조건이 들어오는게 Car객체가 알아야 할 내용이라 생각이 드는데 어떻게 생각하시나요?

makga87 replied:
> [Recommend-Request] Car 객체에 게임의 룰이 들어 가 있습니다. 게임의 룰이 변경 될 경우(예를들어, 특정 값 이상이 아니라, 이하로 변경되었을때) Car객체는 변경이 없고, 차가 움직이는 조건에 대한 클래스가 변경되면 더 좋을 것 같습니다. 오토바이나, 자전거 등 레이싱 가능 한 객체가 늘어났을때 해당 객체들을 수정하는게 아니라, 해당 조건만 수정하면 더 좋을것 같다는 생각이 들었어요.
> 
> 번외로, Car 객체는 CarRaceDifficulty(게임 난이도)를 몰라도 될 것 같습니다. CarRaceDifficulty가 들어오는 순간 해당 Car의 객체는 해당 레이싱 게임에서만 사용 가능한 객체가 된것 같습니다. 어떨때 차가 움직여야한다는 조건이 들어오는게 Car객체가 알아야 할 내용이라 생각이 드는데 어떻게 생각하시나요?

1. 우선 처음 개발 당시에는 Car 객체는 move만 하고, move 조건은 CarRace쪽에서 판단하도록 하였으나, 지난 코드 리뷰에서, 사다리게임 예시로 Car 움직임의 주체이고, 주체가 직접 움직임 여부를 판단하는 방향이 좋겠다라는 리뷰를 받아, 납득이 가는 부분이 있어 위와 같이 수정하였습니다.

2. 게임의 룰이 변경되는 경우는 요구사항에 없어, 우선 위 처럼 개발하였습니다. 

3. "차가 움직이는 조건에 대한 클래스가 변경되면 더 좋을 것 같습니다." 해당 부분은 잘 이해가 안되는데요, 차가 움직이는 조건은 Car 객체에 있는데, Car객체를 변경하라는 말씀이실까요? 아니면 조건 관련된 클래스를 '추가'하는 방향을 말씀하시는 걸까요?

4. "번외로, Car 객체는 CarRaceDifficulty(게임 난이도)를 몰라도 될 것 같습니다.
CarRaceDifficulty가 들어오는 순간 해당 Car의 객체는 해당 레이싱 게임에서만 사용 가능한 객체가 된것 같습니다.
어떨때 차가 움직여야한다는 조건이 들어오는게 Car객체가 알아야 할 내용이라 생각이 드는데 어떻게 생각하시나요?"
이 부분은, 위 1번 내용과 연관이 있어 보입니다. 초기 개발 모델에서는 말씀하신 방향과 마찬가지로, CarRace객체에서 전부 움직임에 대한 조건을 관리하고 있었습니다. 
1번 문의에 대한 토론이 개발자들과 잘 협의가 이뤄진 후, 결정이 되면 1~4번 전부 연관있게 수정이 될 것 같습니다.

HwangSeonHo replied:
저도 1번 이것 때문에 계속 고민 중이라 의견이 궁금하네요.(저는 패키지 구조 만들다보니..)
move는 car뿐만 아니라 다른 곳에서도 쓸 수 있는 기능인데
이걸 car 안으로 넣으면 car가 아닌 곳에서 재사용이 안돼서 move에 대한 코드 중복이 발생하고
car 밖으로 빼자니 티링 생각했을때 basebo처럼 가져다 쓰는 곳이 많아지다가 결합도가 높아져서 독립적으로 사용이 불가능하게 될 것 같아서요

jinyeeun replied:
> > [Recommend-Request] Car 객체에 게임의 룰이 들어 가 있습니다. 게임의 룰이 변경 될 경우(예를들어, 특정 값 이상이 아니라, 이하로 변경되었을때) Car객체는 변경이 없고, 차가 움직이는 조건에 대한 클래스가 변경되면 더 좋을 것 같습니다. 오토바이나, 자전거 등 레이싱 가능 한 객체가 늘어났을때 해당 객체들을 수정하는게 아니라, 해당 조건만 수정하면 더 좋을것 같다는 생각이 들었어요.
> > 번외로, Car 객체는 CarRaceDifficulty(게임 난이도)를 몰라도 될 것 같습니다. CarRaceDifficulty가 들어오는 순간 해당 Car의 객체는 해당 레이싱 게임에서만 사용 가능한 객체가 된것 같습니다. 어떨때 차가 움직여야한다는 조건이 들어오는게 Car객체가 알아야 할 내용이라 생각이 드는데 어떻게 생각하시나요?
> 
> 1. 우선 처음 개발 당시에는 Car 객체는 move만 하고, move 조건은 CarRace쪽에서 판단하도록 하였으나, 지난 코드 리뷰에서, 사다리게임 예시로 Car 움직임의 주체이고, 주체가 직접 움직임 여부를 판단하는 방향이 좋겠다라는 리뷰를 받아, 납득이 가는 부분이 있어 위와 같이 수정하였습니다.
> 2. 게임의 룰이 변경되는 경우는 요구사항에 없어, 우선 위 처럼 개발하였습니다.
> 3. "차가 움직이는 조건에 대한 클래스가 변경되면 더 좋을 것 같습니다." 해당 부분은 잘 이해가 안되는데요, 차가 움직이는 조건은 Car 객체에 있는데, Car객체를 변경하라는 말씀이실까요? 아니면 조건 관련된 클래스를 '추가'하는 방향을 말씀하시는 걸까요?
> 4. "번외로, Car 객체는 CarRaceDifficulty(게임 난이도)를 몰라도 될 것 같습니다.
>    CarRaceDifficulty가 들어오는 순간 해당 Car의 객체는 해당 레이싱 게임에서만 사용 가능한 객체가 된것 같습니다.
>    어떨때 차가 움직여야한다는 조건이 들어오는게 Car객체가 알아야 할 내용이라 생각이 드는데 어떻게 생각하시나요?"
>    이 부분은, 위 1번 내용과 연관이 있어 보입니다. 초기 개발 모델에서는 말씀하신 방향과 마찬가지로, CarRace객체에서 전부 움직임에 대한 조건을 관리하고 있었습니다.
>    1번 문의에 대한 토론이 개발자들과 잘 협의가 이뤄진 후, 결정이 되면 1~4번 전부 연관있게 수정이 될 것 같습니다.

1. Car가 move의 주체가 되어야 하는것은 동의하나, 게임 룰을 Car 객체가 가지고있는것을 말하는 것이였습니다. (둘은 다른 의미로 보입니다.)
2. 게임의 룰이 변경된다는 조건은 없지만, 만약 수정 사항이 생긴다면, Car의 객체가 수정되는게 맞나?라는 의문이 생겨서 질문 드렸습니다.
3. 다양한 방법이 있겠지만,  조건 관련된 클래스를 '추가'하는 방향으로 간다면 1,2번도 자연스레 해결되는 문제라 생각됩니다.

makga87 replied:
> > > [Recommend-Request] Car 객체에 게임의 룰이 들어 가 있습니다. 게임의 룰이 변경 될 경우(예를들어, 특정 값 이상이 아니라, 이하로 변경되었을때) Car객체는 변경이 없고, 차가 움직이는 조건에 대한 클래스가 변경되면 더 좋을 것 같습니다. 오토바이나, 자전거 등 레이싱 가능 한 객체가 늘어났을때 해당 객체들을 수정하는게 아니라, 해당 조건만 수정하면 더 좋을것 같다는 생각이 들었어요.
> > > 번외로, Car 객체는 CarRaceDifficulty(게임 난이도)를 몰라도 될 것 같습니다. CarRaceDifficulty가 들어오는 순간 해당 Car의 객체는 해당 레이싱 게임에서만 사용 가능한 객체가 된것 같습니다. 어떨때 차가 움직여야한다는 조건이 들어오는게 Car객체가 알아야 할 내용이라 생각이 드는데 어떻게 생각하시나요?
> > 
> > 
> > 
> > 1. 우선 처음 개발 당시에는 Car 객체는 move만 하고, move 조건은 CarRace쪽에서 판단하도록 하였으나, 지난 코드 리뷰에서, 사다리게임 예시로 Car 움직임의 주체이고, 주체가 직접 움직임 여부를 판단하는 방향이 좋겠다라는 리뷰를 받아, 납득이 가는 부분이 있어 위와 같이 수정하였습니다.
> > 2. 게임의 룰이 변경되는 경우는 요구사항에 없어, 우선 위 처럼 개발하였습니다.
> > 3. "차가 움직이는 조건에 대한 클래스가 변경되면 더 좋을 것 같습니다." 해당 부분은 잘 이해가 안되는데요, 차가 움직이는 조건은 Car 객체에 있는데, Car객체를 변경하라는 말씀이실까요? 아니면 조건 관련된 클래스를 '추가'하는 방향을 말씀하시는 걸까요?
> > 4. "번외로, Car 객체는 CarRaceDifficulty(게임 난이도)를 몰라도 될 것 같습니다.
> >    CarRaceDifficulty가 들어오는 순간 해당 Car의 객체는 해당 레이싱 게임에서만 사용 가능한 객체가 된것 같습니다.
> >    어떨때 차가 움직여야한다는 조건이 들어오는게 Car객체가 알아야 할 내용이라 생각이 드는데 어떻게 생각하시나요?"
> >    이 부분은, 위 1번 내용과 연관이 있어 보입니다. 초기 개발 모델에서는 말씀하신 방향과 마찬가지로, CarRace객체에서 전부 움직임에 대한 조건을 관리하고 있었습니다.
> >    1번 문의에 대한 토론이 개발자들과 잘 협의가 이뤄진 후, 결정이 되면 1~4번 전부 연관있게 수정이 될 것 같습니다.
> 
> 1. Car가 move의 주체가 되어야 하는것은 동의하나, 게임 룰을 Car 객체가 가지고있는것을 말하는 것이였습니다. (둘은 다른 의미로 보입니다.)
> 2. 게임의 룰이 변경된다는 조건은 없지만, 만약 수정 사항이 생긴다면, Car의 객체가 수정되는게 맞나?라는 의문이 생겨서 질문 드렸습니다.
> 3. 다양한 방법이 있겠지만,  조건 관련된 클래스를 '추가'하는 방향으로 간다면 1,2번도 자연스레 해결되는 문제라 생각됩니다.

1. Car객체가 가지고 있지 않고, 메서드 파라메터로 받고 있는데, 해당 내용도 객체가 가지고 있는것으로 봐야하나요? move가 주체적으로 되어야 한다면, 외부 조건을 판별하기 위해서 메서드를 받아야 한다고 생각이 드는데요, 혹시, 말씀하시고자 하는 부분이, move 조건 자체가 외부로 빠져야 한다는 걸까요? 그럼 move가 주체적으로 일어나지 않는것 아닌가요?

2. 만약 수정사항이 생긴다면 게임의 룰에 영향을 받는 부분은 Car.move이니, 해당 부분이 수정되도 괜찮지 않을까요? 혹시, 제안하고자 하시는 부분이 있다면, 혹시 코드로 설명해주실 수 있으실까요? 글로만 봐선 조금 헷갈려서요

3. 그렇다면 CarRaceDifficulty (난이도)안에다가 넣으면 어떨까요? 이미 난이도 enum이 이동조건과 밀접한 연관이 있어, 별도의 클래스가 추가되는 것 보다요

pjhsk113 replied:
저도 예은님 말씀에 동의하는 부분이 있었어서, 이전 리뷰 내용 중 `전략 패턴`을 활용하면 좋을 것 같다고 말씀드린 적이 있는데요!
이를 활용하면 굉장히 쉽게 해결할 수 있는 문제입니다.

이동 조건에 대한 책임을 MoveStrategy 인터페이스를 구현한 구현체가 가지도록 하면 아주 손쉽게 이동 조건을 변경하거나 추가할 수 있게 됩니다. 또한, 다른 레이싱 주체가 생기더라도 쉽게 이동 조건을 구현할 수 있게됩니다.

```java
public void move(MoveStrategy carMoveStrategy) {
      if (carMoveStrategy.movable()) {
	      this.position++;
      }
}
```

makga87 replied:
두 분 내용이 뭔지 이해했습니다. 수정하겠습니다.

jinyeeun reviewed (commented):
대체로 코드가 깔끔하고, 읽기 편안했습니다.
한가지 제안을 해보았는데, 확인 부탁드립니다.

makga87 committed:
[FIX] https://github.com/Effective-Tilinkle/racingcar-with-java/pull/4#discussion_r1096891890

jinyeeun reviewed (approved):
👍
